numpy/core/src/multiarray/methods.c

```diff
@@ -2598,45 +2598,24 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
 
     if (!PyArray_CanCastArrayTo(self, dtype, NPY_SAME_KIND_CASTING) &&
             !(PyArray_TYPE(self) == NPY_OBJECT)) {
-        PyObject *err, *msg_part;
+        PyObject *descr = (PyObject*)PyArray_DESCR(self);
+
+        PyErr_Format(PyExc_TypeError, "Unable to convert %R to complex", descr);
         Py_DECREF(dtype);
-        err = PyString_FromString("unable to convert ");
-        if (err == NULL) {
-            return NULL;
-        }
-        msg_part = PyObject_Repr((PyObject*)PyArray_DESCR(self));
-        if (msg_part == NULL) {
-            Py_DECREF(err);
-            return NULL;
-        }
-        PyString_ConcatAndDel(&err, msg_part);
-        if (err == NULL) {
-            return NULL;
-        }
-        msg_part = PyString_FromString(", to complex.");
-        if (msg_part == NULL) {
-            Py_DECREF(err);
-            return NULL;
-        }
-        PyString_ConcatAndDel(&err, msg_part);
-        if (err == NULL) {
-            return NULL;
-        }
-        PyErr_SetObject(PyExc_TypeError, err);
-        Py_DECREF(err);
         return NULL;
     }
 
     if (PyArray_TYPE(self) == NPY_OBJECT) {
         /* let python try calling __complex__ on the object. */
         PyObject *args, *res;
-        Py_DECREF(dtype);
```

Comment:
Why move this?

Reply:
Because we usually do clean just before returns. I find the code reads more logically this way as it wasn't immediately obvious why the reference was decremented where it was, you had to read through the code to see that it wasn't used and that there was a return.

Reply:
Well as written, this now leaks `dtype` from line 2636 ;)

Reply:
Fixed. The whole function can be rewritten but it doesn't look any prettier when I do.